web/app/Models/Interaction.php

```diff
@@ -61,73 +61,148 @@ public function target() : MorphTo
+     * @param array|null $metadata
+     * @return bool
+     */
+    public static function isInvalid($setter, $relation, $target, ?array $metadata = []) : bool
+    {
+        return !(
+            is_subclass_of($setter, 'Illuminate\Database\Eloquent\Model')
+            && InteractionType::isValue($relation)
+            && is_subclass_of($target, 'Illuminate\Database\Eloquent\Model')
+            && is_array($metadata)
+        );
+    }
+
     /**
      * Create and persist a new Interaction object.
      * @param $setter
      * @param int $relation
      * @param $target
      * @param array|null $metadata
-     * @return Interaction
+     * @return bool
      */
-    public static function add($setter, int $relation, $target, ?array $metadata = []) : ?Interaction
+    public static function create($setter, int $relation, $target, ?array $metadata = []) : bool
     {
-
-        /**
-         * We have to do some validation here as we're accepting generics.
-         */
-        if(is_subclass_of($setter, 'Illuminate\Database\Eloquent\Model')
-            && InteractionType::isValue($relation)
-            && is_subclass_of($target, 'Illuminate\Database\Eloquent\Model'))
+        if(self::isInvalid($setter, $relation, $target, $metadata))
         {
-            $interaction = new Interaction(
-                [
-                    'setter_type' => get_class($setter),
-                    'setter_id' => $setter->id,
-                    'interaction_type' => $relation,
-                    'target_type' => get_class($target),
-                    'target_id' => $target->id,
-                    'metadata' => $metadata
-                ]
-            );
+            Log::error("Interaction::add was given an invalid set of params. 
+            Setter: $setter Relation: $relation Target: $target Metadata: $metadata");
+            abort(500);
+            return false;
+        }
 
-            try {
-                $interaction->save();
-                return $interaction;
-            }
-            catch(QueryException $e) {
-                /** Postgres unique constraint violation: */
-                if($e->errorInfo[0] == 23505) {
-                    /**
-                     * We'll want to throw something here that a controller can
-                     * catch and return data to the user. Pending API work.
-                     */
-                    return null;
-                }
+        $interaction = new Interaction(
+            [
+                'setter_type' => get_class($setter),
+                'setter_id' => $setter->id,
+                'interaction_type' => $relation,
+                'target_type' => get_class($target),
+                'target_id' => $target->id,
+                'metadata' => $metadata
+            ]
+        );
+
+        try {
+            return $interaction->save();
+        }
+        catch(QueryException $e) {
+            /** Postgres unique constraint violation: */
```

Comment:
Is there a way we can wrap `QueryExtension`? Having to remember magic postgres error codes doesn't seem like great ergonomics long term.

Reply:
I believe there is already a specific exception for unique constraint violations, I'll try and find it.